src/wasm/wasm-binary.cpp

```diff
@@ -516,40 +522,66 @@ uint32_t WasmBinaryWriter::getTypeIndex(HeapType type) const {
   return it->second;
 }
 
-void WasmBinaryWriter::writeFunctionTableDeclaration() {
-  if (!wasm->table.exists || wasm->table.imported()) {
+void WasmBinaryWriter::writeTableDeclarations() {
+  if (importInfo->getNumDefinedTables() == 0) {
+    // std::cerr << std::endl << "(WasmBinaryWriter::writeTableDeclarations) No
+    // defined tables found. skipping" << std::endl;
     return;
   }
-  BYN_TRACE("== writeFunctionTableDeclaration\n");
+  BYN_TRACE("== writeTableDeclarations\n");
   auto start = startSection(BinaryConsts::Section::Table);
-  o << U32LEB(1); // Declare 1 table.
-  o << S32LEB(BinaryConsts::EncodedType::funcref);
-  writeResizableLimits(wasm->table.initial,
-                       wasm->table.max,
-                       wasm->table.hasMax(),
-                       /*shared=*/false,
-                       /*is64*/ false);
+  auto num = importInfo->getNumDefinedTables();
+  o << U32LEB(num);
+  ModuleUtils::iterDefinedTables(*wasm, [&](Table* table) {
+    o << S32LEB(BinaryConsts::EncodedType::funcref);
+    writeResizableLimits(table->initial,
+                         table->max,
+                         table->hasMax(),
+                         /*shared=*/false,
+                         /*is64*/ false);
+  });
   finishSection(start);
 }
 
 void WasmBinaryWriter::writeTableElements() {
-  if (!wasm->table.exists || wasm->table.segments.size() == 0) {
+  if (wasm->tables.empty()) {
+    return;
+  }
```

Comment:
In the case where there is at least one table but all the tables are empty, this early return won't happen and we will write out a useless table element section. I think it would be worth iterating through the tables and checking whether they are all empty, probably like this:
```suggestion
  if (std::all_of(wasm->tables.begin(), wasm->tables.end(), [&](auto& table) {
        return std::all_of(table->segments.begin(), table->segments.end(), [&] (auto& segment) {
          return segment.size() == 0;
        })
      })) {
    // There aren't any table segments, so skip this section.
    return;
  }
```

Reply:
Right, this early condition was unnecessary because I'm counting segments of non-empty tables right after and return if the total count is zero. So I just removed it.